src/naf-dialog-adapter.js

```diff
@@ -143,26 +192,159 @@ export default class DialogAdapter {
     }
   }
 
+  /**
+   * This applies to the ICE restart method below.
+   * Why do we need new ICE servers? In case the ICE fails we need to re-negotiate new candidates
+   * but most probably our TURN credentials have expired so we need to request new credentials and
+   * update the TURN servers to be able to gather new candidates.
+   * Firefox doesn't support hot updating ICE servers through mediasoup's updateIceServers (which
+   * internally uses setConfiguration):
+   * https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/setConfiguration
+   * So we recreate the transports/consumer/producers through reconnect to force setting
+   * new ICE servers for both send and receive transports.
+   */
+
   /**
    * Restart ICE in the underlying send peerconnection.
    */
-  async restartSendICE() {
+  async restartSendICE(force = true) {
     if (!this._sendTransport?._closed) {
-      this.emitRTCEvent("log", "RTC", () => `Restarting send transport ICE`);
-      this._sendTransport.restartIce({ iceParameters: this._sendIceParameters });
+      // Renegotiate ICE in case ICE state is failed, disconnected. The last two conditions cover the cases where the
+      // Send transport has been create but for whatever reason is stuck in new state and doesn't gather candidates.
+      if (
+        force ||
+        this._sendTransport.connectionState === "failed" ||
+        (!isFirefox && this._sendTransport.connectionState === "disconnected") ||
+        (this._sendTransport.connectionState === "new" && this._lastSendConnectionState === "failed")
+      ) {
+        this.emitRTCEvent("log", "RTC", () => `Restarting send transport ICE`);
+        if (this._protoo.connected) {
+          if (isFirefox) {
+            this.reconnect();
+          } else if (!this._sendTransport.closed) {
+            const { host, port, turn } = await window.APP.hubChannel.getHost();
+            const iceServers = this.getIceServers(host, port, turn);
+            await this._sendTransport.updateIceServers({ iceServers });
+            const iceParameters = await this._protoo.request("restartIce", { transportId: this._sendTransport.id });
+            await this._sendTransport.restartIce({ iceParameters });
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Checks the Send Transport ICE status and restarts it in case is in failed or disconnected states.
+   * This is called by the Send Transport "connectionstatechange" event listener.
+   * @param {boolean} connectionState The transport connnection state (ICE connection state)
+   */
+  checkSendIceStatus(connectionState) {
+    // If the ICE connection state failed we force an ICE negotiation
+    if (connectionState === "failed" || connectionState === "disconnected") {
+      this.restartSendICE(false);
+    } else if (connectionState === "connected") {
+      this._iceReconnectSendTaskId && clearTimeout(this._iceReconnectSendTaskId);
```

Comment:
Is this a new javascript syntax for a simple if statement? That's awesome!

Reply:
You can see it that way, it's just a slightly more concise and idiomatic way of doing a condition check and save some lines of code. Same as the safe call operator (?.)